### PR TITLE
SW-4410 Modify planting seasons

### PR DESCRIPTION
**Pull request by sgrimm**
Implement the write path for planting seasons. They can be included at planting
site creation time and can be updated on existing sites.

### CODE REVIEW

**Comment by sgrimm** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1492** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1492?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #1493** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1493?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1494** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1494?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
        * **PR #1495** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1495?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #1496** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #1497** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1497?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #1498** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1498?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #1499** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1499?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #1500** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1500?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                    * **PR #1501** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1501?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1494?utm_source=stack-comment).

File: src/main/kotlin/com/terraformation/backend/tracking/model/PlantingSeasonModel.kt [hunk 1]
```diff
@@ -12,3 +12,35 @@ data class ExistingPlantingSeasonModel(
...
+  constructor(
+      existingModel: ExistingPlantingSeasonModel
+  ) : this(
+      endDate = existingModel.endDate,
+      id = existingModel.id,
+      startDate = existingModel.startDate,
+  )
+
+  fun validate(currentDate: LocalDate) {
+    if (endDate < startDate.plusDays(28)) {
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/tracking/model/PlantingSeasonModel.kt:30]: do we want another check for endDate <= startDate? (i understand there's a db schema constraint anyways)

**Review by karthikbtf — APPROVED** [review]
Seems clear to me, just one question but not so important.

File: src/main/kotlin/com/terraformation/backend/tracking/db/PlantingSiteStore.kt [hunk 2]
```diff
@@ -813,4 +921,37 @@ class PlantingSiteStore(
         ?.value1()
         ?: throw IllegalStateException("Could not query zone's permanent clusters")
   }
+
+  private fun validatePlantingSeasons(
```
Comment (nickgraz) [inline src/main/kotlin/com/terraformation/backend/tracking/db/PlantingSiteStore.kt:934]: ~Perhaps the planting season length requirements should go here?~ Never mind, found them!

**Review by nickgraz — APPROVED** [review]

File: src/main/kotlin/com/terraformation/backend/tracking/model/PlantingSeasonModel.kt (hunk 1 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/tracking/model/PlantingSeasonModel.kt:30]: You mean to return a different error? The existing check should catch that case (if end date is before start date, then it'll also be before start date + 28 days).

Comment (karthikbtf) [reply src/main/kotlin/com/terraformation/backend/tracking/model/PlantingSeasonModel.kt:30]: Of course, my mistake. This is clear, thanks 🙏

**Comment by sgrimm** [comment]
### Merge activity

* **Nov 27, 5:00 PM**: @sgrimm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1494).
* **Nov 27, 5:01 PM**: [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1494) rebased this pull request as part of a merge.
* **Nov 27, 5:02 PM**: @sgrimm merged this pull request with [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1494).

**Pull request merged**